Apple: Improve Apple embedded export and debugging

stuartcarnie opened the pull request:
# Summary

- Improve Apple unified logging system (`os_log`) for better debugging on iOS/macOS/visionOS

https://github.com/user-attachments/assets/fd2794c8-3f9a-4f42-ac1b-929706018468

> [!NOTE]
>
> In the above video, which demonstrates running a game on a physical iPhone, we can use the bundle identifier to examine on-device logs in real time, and filter by specific categories. This works even when the game is not attached to the debugger.

- Fix remote filesystem bug for Apple embedded platform development

<p align="center">
<img src="https://github.com/user-attachments/assets/2d2d0fb8-c0f7-4698-865d-fe720dfaee9a" width="30%" />
</p>

- Fix `devicectl` device discovery, to correctly separate devices by platform type (iOS, visionOS, and later tvOS)

<p align="center">
<img src="https://github.com/user-attachments/assets/0255b0d4-4a78-4c9c-afe6-64b8442f9f6b" width="50%" />
</p>

- Improve reliability of direct device deployment when compiling and installing apps to Apple embedded devices like an iPhone or Vision Pro

Implemented an alternative that uses `execute_with_pipe` to reliably run `xcodebuild` and `xcrun` tooling

- Add Foundation string helpers for efficient conversion between NSString and Godot String types
  - These are only used in code that was added for this PR, but a separate PR could be added to use more widely

stuartcarnie commented on core/debugger/remote_debugger_peer.cpp:
```diff
@@ -96,11 +98,13 @@ void RemoteDebuggerPeerTCP::_write_out() {
 		if (out_left <= 0) {
-			if (out_queue.is_empty()) {
-				break; // Nothing left to send
-			}
 			mutex.lock();
-			Variant var = out_queue.front()->get();
+			List<Array>::Element *E = out_queue.front();
+			if (!E) {
+				mutex.unlock();
+				break;
+			}
+			Variant var = E->get();
```
This was a data-race, as it was reading the `is_empty` outside the mutex. Simplify the code to just pop the front element and check for `nullptr`

stuartcarnie commented on drivers/apple/foundation_helpers.mm:
```diff
@@ -0,0 +1,79 @@
+namespace conv {
+
+NSString *to_nsstring(const String &p_str) {
+	return [[NSString alloc] initWithBytes:(const void *)p_str.ptr()
+									length:p_str.length() * sizeof(char32_t)
+								  encoding:NSUTF32LittleEndianStringEncoding];
```
In-place conversion of UTF-32 strings to NSString saves allocating an intermediate `CharString<char>`.

stuartcarnie replied:
I did that originally, but for some reason the category wasn’t loading, and throwing a “selector not found” exception. 

There is no reason it shouldn’t work, so I will try again.

stuartcarnie reviewed (commented):
Comments about changes

bruvzg commented on platform/macos/os_macos.mm:
```diff
@@ -758,6 +760,113 @@
+		ERR_PRINT(vformat("launch failed: %s", conv::to_string(err.localizedDescription)));
+		return FAILED;
+	}
+	[task waitUntilExit];
```
Generic Unix implementation is not waiting for completion, but instead reading from the pipe as data arrives. Same with the Windows implementation. So it's not identical behavior.

bruvzg replied:
> It is waiting, as that is what [pclose](https://www.man7.org/linux/man-pages/man3/pclose.3p.html) does

It's waiting only after reading is done, not before it. So if it's called from a thread, another thread can get partially read pipe data while the process is still running. But I'm not sure if it's actually used anywhere. Exposed GDScript wrapper only return when pipe is closed, so it can be used only internally.

bruvzg reviewed (commented):
Not sure what's the point of adding `NSTask` implementation to the `execute`.

stuartcarnie commented on drivers/apple/foundation_helpers.mm:
```diff
@@ -0,0 +1,79 @@
+				return String::ascii(Span(p, CFStringGetLength(str)));
+			case kCFStringEncodingUTF8:
+				return String::utf8(p);
+			case kCFStringEncodingUTF32LE:
```
That is what I tried first and oddly enough, `kCFStringEncodingUTF32` is different from `kCFStringEncodingUTF32LE`, and didn't work. I would have thought `kCFStringEncodingUTF32` was defined to match the endianness of the platform (which it probably is internally), but they are different constants.

stuartcarnie commented on platform/ios/export/export_plugin.h:
```diff
@@ -36,8 +36,8 @@ class EditorExportPlatformIOS : public EditorExportPlatformAppleEmbedded {
 	virtual String get_platform_name() const override { return "ios"; }
-
 	virtual String get_sdk_name() const override { return "iphoneos"; }
+	virtual String get_device_type() const override { return "iPhone"; }
```
Good catch. I'll use `platform` instead.

stuartcarnie commented:
> If it's unreliable, the Unix implementation probably should be fixed (or replaced) instead. Is `execute_with_pipe` working better?

I didn't try that, but it is quite different, as it relies on the caller to consume the data via the pipes.

I refactored the UNIX implementation of `execute` to use `feof` and `fread` to read in chunks. Just like the prior implementation, the pipe returned by `popen` is returning an error at various stages of compilation:

![CleanShot 2025-06-17 at 07 20 42@2x](https://github.com/user-attachments/assets/a7ce0a4b-33db-4858-a3d3-a1e9c96dc3bb)

In this example, you can see that it accumulated data up to `GatheringProvisioningInputs` before `f` returns an error. I think this is the same thing that is happening with `fgets`, which no doubt is checking for errors. 

I don't know why, but using `NSTask` reliably runs to completion. `popen` spawns the process using the default shell, and perhaps that is causing issues with `xcodebuild`.

stuartcarnie commented on platform/macos/os_macos.mm:
```diff
@@ -1023,8 +1132,8 @@
-
 	Vector<Logger *> loggers;
+	loggers.push_back(memnew(OsLogLogger(NSBundle.mainBundle.bundleIdentifier.UTF8String)));
 	loggers.push_back(memnew(MacOSTerminalLogger));
```
I'll remove this and the old Apple Embedded one too

rsanchezsaez reviewed (approved):
Great PR! Left a few minor comments.

LGTM from my end, pending agreement about whether `NSTask` is the best approach here: https://github.com/godotengine/godot/pull/107574#issuecomment-2978197423

(I had also observed the issue @stuartcarnie reported about the exporter not seeing the full output of `xcodebuild` when trying to export an IPA, but I don't know what the best solution is.)

stuartcarnie commented on editor/export/editor_export_platform_apple_embedded.h:
```diff
@@ -91,6 +91,13 @@ class EditorExportPlatformAppleEmbedded : public EditorExportPlatform {
 	void _update_preset_status();
+
+protected:
+	void _start_thread() {
```
Yes, agree – I'll change that.

stuartcarnie commented:
@bruvzg I've updated the PR as follows:

- Removed the `OS_MacOS::execute` override using `NSTask`
- Introduced an `_execute` method to `EditorExportPlatformAppleEmbedded` to use `execute_with_pipe`
  - Now reliably executes build and deploys to physical device
- Moved debugger issues to #107643

## `execute_with_pipe` API

This API feels a bit clunky and needs some fixes / improvements:

- Never cleans up old processes from the process map:

  https://github.com/godotengine/godot/blob/24494d840e06402c26c03db36695b63c5322d624/drivers/unix/os_unix.cpp#L800-L803

- Doesn't proved a way to wait for a process to complete (you must poll)

For 4.6, I think I'll submit a PR to improve `OS_Unix::execute` to use the same approach, rather than `popen`.

bruvzg commented:
Apart from logger changes (which should be safe as well) and `foundation_helpers` (seems like it's not used in the current implementation, so can be moved to separate cleanup PR for 4.6, with full performance tests and replacing all `NSString` conversion cases if it worth it), it's bugfix for iOS/visionOS export and definitely is OK for 4.5.

AThousandShips commented on drivers/apple/os_log_logger.cpp:
```diff
@@ -0,0 +1,127 @@
+#include "core/string/print_string.h"
+#include "core/string/ustring.h"
+
+#include <os/log.h>
```
```suggestion
```
Already included

stuartcarnie committed:
Apple: Improve reliability of Apple Embedded build and deployment

owlnewworlds commented:
Sorry if my comment opens the pull again but I'm dying to find informations of if this PR will change anything about the process of iOS export / Xcode build? It's still a real enormous hurdle to export a game to iOS to upload it on Appstore and it looks like the PR is changing a lot of stuff regarding export.

Did the process change? Is it more straightforward now?

owlnewworlds commented:
> > Sorry if my comment opens the pull again but I'm dying to find informations of if this PR will change anything about the process of iOS export / Xcode build? It's still a real enormous hurdle to export a game to iOS to upload it on Appstore and it looks like the PR is changing a lot of stuff regarding export.
> > Did the process change? Is it more straightforward now?
> 
> This PR improves the reliability of running the project generation and build tools.
> 
> @owlnewworlds could you provide specific information about the issues you are running into?

I think most of the issues I'm running into will be solved by this PR, I will test the Godot 4.5 beta 1 soon.
BUT I have also one major issue with the following situation (it may be off topic though) :

I'm making updates in the Godot iOS official plugins repository because I need more functions and I was able to build the plugins and add them in my Godot game project. But when I export the project and build it / run it with XCode, it seems it run the old version of the plugins without my changes. I was never able to figure out what could happen...

owlnewworlds commented:
Is [this kind of issue](https://github.com/godotengine/godot/issues/57195) will be solved with the PR?
